### ISSUE TITLE
"dart pub global activate" causes to "pub: command not found"

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/sdk#42530]
On the latest version of Dart, `dart pub global activate` doesn't seem to work.

Example:

```dart
// bin/my_echo.dart
void main(List<String> args) {
  for (final str in args) {
    print(str);
  }
}
```

```yaml
// pubspec.yaml
name: my_echo

executables:
  my_echo:
```

Then run:

```dart
dart pub global activate -s path .
my_echo "hello world"
```

This leads to:

```dart
/Users/remirousselet/.pub-cache/bin/my_echo: line 7: pub: command not found
```

Dart version: 

```dart
Dart SDK version: 2.9.0-19.0.dev.flutter-f91547d6dd (be) (Sat Jun 27 19:14:41 2020 +0000) on "macos_x64"
```

**Comment by bkonyi** [comment]
I'm unable to reproduce on Linux or OSX at f91547d6dd or HEAD running directly from the Dart SDK. I'm also unable to reproduce on the latest Flutter beta channel on Linux so I'm going to close this for now. 

If you're still running into this, it sounds like you might not have `pub` on your path.

**Comment by rrousselGit** [comment]
@bkonyi  the problem is that `pub` as a command line was removed. Instead we have `dart pub`.

But using `dart pub global activate` still tries to call the legacy `pub` cli.

**Comment by bkonyi** [comment]
As far as I know we haven't removed any of the legacy scripts for tooling, especially since this CLI isn't published on stable yet. We'll do eventually remove the standalone tool scripts, but not without an adequate deprecation warning. Building from HEAD still spits out a `pub` script when building the SDK.

**Comment by rrousselGit** [comment]
That is not the case with the Flutter SDK

It ships only `dart` and `flutter`, as seen in its `bin/` folder.

**Comment by bkonyi** [comment]
Ah, I see the problem here. It sounds like you have `flutter/bin` on your path but not `flutter/bin/cache/dart-sdk/bin`, which is where all the Dart tooling scripts live. We originally didn't expect people to be using the Dart SDK packaged with Flutter directly, so it's kind of buried in the artifacts (I think this is something changing in the future).

**Comment by rrousselGit** [comment]
I see thanks! 

Would it make sense to change the generated .pub-cache/bin/my_binary file to generate `dart pub x` instead of `pub x` instead?

That would fix the issue without exposing the raw pub cli

**Comment by bkonyi** [comment]
No problem!

> Would it make sense to change the generated .pub-cache/bin/my_binary file to generate dart pub x instead of pub x instead?

That's not a bad idea. Thoughts @jwren?

**Comment by jwren** [comment]
No objection from me, but I am not familiar with all of the consequences of adding content to the generated `.pub-cache`.  @bkonyi do you know who manages the generation of those files?

**Comment by bkonyi** [comment]
Maybe @jonasfj would know?

**Comment by jonasfj** [comment]
`pub global activate` manages the contents of `$PUB_CACHE/bin/`.

We are having a similar issue in https://github.com/dart-lang/pub/issues/2593

TL;DR: The contents of `$PUB_CACHE/bin/` is generated when `pub global activate <package>` runs. But we might need to find a way to update scripts... especially given the changes happening with:
 * how we do snapshots (https://github.com/dart-lang/pub/issues/2593), and,
 * `dart pub` replacing `pub` (in particular, `pub` not shipping with Flutter, while `dart pub` does).

/cc, @sigurdm we should talk about this tomorrow.

**Comment by bradyt** [comment]
Just a reminder, I think flutter installation documentation leaves it easy to end up searching for a solution. It seems to suggest using the `flutter/bin/dart`, as read here: <https://flutter.dev/docs/get-started/install/macos>.

>  Note: As of Flutter’s 1.19.0 dev release, the Flutter SDK contains the dart command alongside the flutter command so that you can more easily run Dart command-line programs.
> ...
> After updating your shell for the change to take effect, running the which or where command again should show that the flutter and dart commands now come from the same directory.

But still, the solution above worked for me, adding `flutter/bin/cache/dart-sdk/bin` to PATH.

**Comment by bkonyi** [comment]
Is this still an issue? Pub is embedded in the CLI now, so there shouldn't be any issues with paths.

**Comment by Fox32** [comment]
> Is this still an issue? Pub is embedded in the CLI now, so there shouldn't be any issues with paths.

Just run into this two days ago on `Flutter (Channel stable, 2.0.1, on Linux, locale C.UTF-8)`. Had to rewrite the generated file to use `dart pub` instead.

**Comment by bkonyi** [comment]
@Fox32 what generated file are you referring to?

**Comment by Fox32** [comment]
The file inside my `/.pub-cache/bin/` that is generated by `pub global activate` is referencing `pub` not `dart pub`. For example, if I run `dart pub global activate pubviz` it generates a file like this:

```shell
#!/usr/bin/env sh
# This file was created by pub v2.12.0.
# Package: pubviz
# Version: 2.8.2+1
# Executable: pubviz
# Script: pubviz
if [ -f /home/oliver/.pub-cache/global_packages/pubviz/bin/pubviz.dart-2.12.0.snapshot ]; then
  dart "/home/oliver/.pub-cache/global_packages/pubviz/bin/pubviz.dart-2.12.0.snapshot" "$@"
  # The VM exits with code 253 if the snapshot version is out-of-date.
  # If it is, we need to delete it and run "pub global" manually.
  exit_code=$?
  if [ $exit_code != 253 ]; then
    exit $exit_code
  fi
  pub global run pubviz:pubviz "$@"
else
  pub global run pubviz:pubviz "$@"
fi
```

Actually using `pubviz` works because a snapshot is generated and run using `dart` in this case. However note, that the other lines are still using `pub`. 

If I install a command from a local dirctopry no snapshot is used. For example: `dart pub global activate --source path mycommand` results in:

```shell
#!/usr/bin/env sh
# This file was created by pub v2.12.0.
# Package: mono
# Version: 0.0.0
# Executable: mono
# Script: mono
pub global run mycommand:mycommand "$@"
``` 

This file is not usable if you don't have a full SDK with pub on the path. You have to change it to `dart pub` to run it.

**Comment by bkonyi** [comment]
> The file inside my `/.pub-cache/bin/` that is generated by `pub global activate` is referencing `pub` not `dart pub`. For example, if I run `dart pub global activate pubviz` it generates a file like this:
> 
> ```shell
> #!/usr/bin/env sh
> # This file was created by pub v2.12.0.
> # Package: pubviz
> # Version: 2.8.2+1
> # Executable: pubviz
> # Script: pubviz
> if [ -f /home/oliver/.pub-cache/global_packages/pubviz/bin/pubviz.dart-2.12.0.snapshot ]; then
>   dart "/home/oliver/.pub-cache/global_packages/pubviz/bin/pubviz.dart-2.12.0.snapshot" "$@"
>   # The VM exits with code 253 if the snapshot version is out-of-date.
>   # If it is, we need to delete it and run "pub global" manually.
>   exit_code=$?
>   if [ $exit_code != 253 ]; then
>     exit $exit_code
>   fi
>   pub global run pubviz:pubviz "$@"
> else
>   pub global run pubviz:pubviz "$@"
> fi
> ```
> 
> Actually using `pubviz` works because a snapshot is generated and run using `dart` in this case. However note, that the other lines are still using `pub`.
> 
> If I install a command from a local directory no snapshot is used. For example: `dart pub global activate --source path mycommand` results in:
> 
> ```shell
> #!/usr/bin/env sh
> # This file was created by pub v2.12.0.
> # Package: mycommand
> # Version: 0.0.0
> # Executable: mycommand
> # Script: mycommand
> pub global run mycommand:mycommand "$@"
> ```
> 
> This file is not usable if you don't have a full SDK with pub on the path. You have to change it to `dart pub` to run it.

Thanks for the details! @jonasfj, can you take a look?

**Comment by longtimedeveloper** [comment]
**Temporary Workaround for Windows Developers**

After running, `dart pub global activate --executable=fonts2yaml --source path . ` I copy the created batch file in the `\AppData\Local\Pub\Cache\bin` directory to another directory in my path.  

I edit the created batch file, adding `dart` to the beginning of the last line so that it now reads, `dart pub global run fonts2yaml:fonts2yaml %*`

**Important:** The other folder is in my path BEFORE the Flutter/Dart SDK entry so that my batch file executes and not the broken batch file.

Problem solved, but not sustainable.

@bkonyi @jonasfj thank you for sorting this out and have a great day!

Karl

**Comment by wonderkidshihab** [comment]
any solution to this problem yet? I am getting this error while installing serverpod.

Unhandled exception:
ProcessException: The system cannot find the file specified.

  Command: which dart
#0      _ProcessImpl._start (dart:io-patch/process_patch.dart:401:33)
#1      Process.start (dart:io-patch/process_patch.dart:38:20)
#2      _runNonInteractiveProcess (dart:io-patch/process_patch.dart:578:18)
#3      Process.run (dart:io-patch/process_patch.dart:49:12)
#4      CommandLineTools.existsCommand (file:///C:/Users/Dell/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/serverpod_cli-0.9.4/bin/create/command_line_tools.dart:19:32)
#5      main (file:///C:/Users/Dell/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/serverpod_cli-0.9.4/bin/serverpod_cli.dart:35:31)
#6      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:32)
#7      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)
Unhandled exception:
ProcessException: The system cannot find the file specified.

  Command: which dart
#0      _ProcessImpl._start (dart:io-patch/process_patch.dart:401:33)
#1      Process.start (dart:io-patch/process_patch.dart:38:20)
#2      _runNonInteractiveProcess (dart:io-patch/process_patch.dart:578:18)
#3      Process.run (dart:io-patch/process_patch.dart:49:12)
#4      CommandLineTools.existsCommand (file:///C:/Users/Dell/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/serverpod_cli-0.9.4/bin/create/command_line_tools.dart:19:32)
#5      main (file:///C:/Users/Dell/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/serverpod_cli-0.9.4/bin/serverpod_cli.dart:35:31)
#6      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:32)
#7      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)

**Comment by longtimedeveloper** [comment]
@wonderkidshihab see my solution above. Hope this helps!
